content/recipes/authoring/kata-snippets/full-tests.md

```diff
@@ -7,4 +7,59 @@ next: /recipes/authoring/kata-snippets/preloaded/
+- **Make debugging of random tests easy.** Sometimes, when random tests fail, they are very difficult to debug because input is very large and cannot be easily printed. If possible, Try to compose your test suite in such a way that when test fail, they fail on some small, easy, debuggable input, before they fail on a large one, which is difficult to print. 
+
+## Performance Tests
+
+Some kata require solutions to be fast, or to fall into some performance category. For example, author might want to accept only solutions of complexity `O(n)`, and reject `O(n^2)` ones. Building such test suite is not always easy!
+
+- **Performance tests can be implemented in terms of random tests.** Usually, it's not reqired to have separate sections for random tests and performance tests. Testing of performance can be driven by size of random inputs. But still, to make debugging easier, it can be worth in some cases to have a set of separate random tests with small inputs.
+- If difficulty of a tested kata is related to the size of the input, it's usually **better to have a few tests with large input, that thousands of tests with small input.** 200 tests with very large numbers or arrays is usually better than 10_000 tests with small ones (but see remarks on size of inputs above).
+- **Difference beetwen accepted and rejected solutions should be easy to spot.** Ideally, tests for accepted solutions should complete very fast, and tests for rejected solutions should take very long. Otherwise, you risk that solutions with valid complexity characteristics will time out, and users will be frustrated looking for micro-optimizations. This approach usually calls for very large inputs, so be careful! 
+- **Performance tests should be consistent between runs.** It should not happen that one and the same solution sometimes passes, and sometimes fails, depending on randomized inputs.
+
+## Tests of Additional Restrictions
+
+Kata may have some additional restrictions, not related to the input and output of the tested function. For example, code golf kata usually call for solutions shorter than X characters, or some kata require the task to be realized without using some class Y. Every such requirement has to be tested explicitly, but authors should be aware thet it's not always easy (or even possible):
+
+- If user solution has to be inspected as text, it can be found in file `/home/codewarrior/solution.txt`.
+- Restrictions in form of _"Do not use X"_ or _"You have to use Y"_  (for example _"do not use BigInteger"_, or _"only recursive solutions are allowed"_) should not be used, because they are very difficult (sometimes even impossible) to be enforced reliably. See [this article(TODO: write the article)]() for details.
```

Comment:
```suggestion
- Restrictions of the form "do not use X" or "you must use Y" are seldom a good idea since they are extremely difficult (or downright infeasible) to reliably enforce. See [this article (TODO: write the article)]() for details.
```